homeassistant/components/proxmoxve/coordinator.py

```diff
@@ -0,0 +1,137 @@
+"""DataUpdateCoordinators for the Proxmox VE integration."""
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import Any
+
+from proxmoxer import AuthenticationError, ProxmoxAPI
+from proxmoxer.core import ResourceException
+from requests.exceptions import ConnectTimeout, SSLError
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import LOGGER, UPDATE_INTERVAL
+from .models import ProxmoxVMData
+
+
+class ProxmoxCoordinator(DataUpdateCoordinator[ProxmoxVMData]):
+    """Proxmox VE data update coordinator."""
+
+
+class ProxmoxQEMUCoordinator(ProxmoxCoordinator):
```

Comment:
There is a lot of duplicate code here. Some of it should be moved to the base `ProxmoxCoordinator`.
It should have:
- an `__init__` method to pass and hold hass and proxmox (maybe node_name)
- an abstract (sync) method `_poll_api` that is overriden in the other coordinators
- `_async_update_data` can then also be shared in the base coordinator

In the future, when you add properties, you can create a `parse_result` abstract method to parse the resulting dict but for now that can stay in the base `_async_update_data` method.

The child coordinators will simply have:
- an `__init__` method to set `qemu_id`/`container_id`
- a `_poll_api` method which simply returns either
  - `self.proxmox.nodes(self.node_name).qemu(self.qemu_id).status.current.get()`
  - or `self.proxmox.nodes(self.node_name).lxc(self.container_id).status.current.get()`

Reply:
But I have a doubt...

The coordinator will have to deal with what will be returned, at that moment all coordinators will return `ProxmoxVMData`, but the expectation is that in the future we will have others for example `ProxmoxNodeData` and `ProxmoxStorageData` and in the future the return of the LXC coordinator may be different from ` ProxmoxVMData` depending on added sensors.

That's why I thought of doing all the data manipulation in the child coordinators.

If the child coordinator just returns `self.proxmox.nodes(self.node_name).qemu(self.qemu_id).status.current.get()` or `self.proxmox.nodes(self.node_name).lxc(self. container_id).status.current.get()` I will have problems in the future to handle data from Nodes and Storage or I would have a very large code in the parent coordinator to evaluate and handle each case.

I want to understand better so that in the future I don't undo what is defined here...

Reply:
In the future, when you add properties, you can create a parse_result abstract method to parse the resulting dict but for now that can stay in the base _async_update_data method.